Is it possible to test Variable Validations?

I am using TFTest to help develop a module and we have some strict variable validation rules. Is it possible to test these out with TFTest?

We are looking to make sure that either the plan fails due to validation error, or that the error message is returned in the output

Currently we are just seeing "plan did not yield an output"

It seems to be working fine for me. Try the following.

```hcl
# main.tf
variable "foo" {
  type = string
  validation {
    condition     = length(var.foo) > 3
    error_message = "Foo neds to be at least 3 characters."
  }
}
```

```python
# test-foo.py
import tftest


def test_plan_fail():
  tf = tftest.TerraformTest('.')
  tf.setup(upgrade=True)
  return tf.plan(output=True, tf_vars={'foo': 'bar'})


def test_plan_success():
  tf = tftest.TerraformTest('.')
  tf.setup(upgrade=True)
  return tf.plan(output=False, tf_vars={'foo': 'spam'})
```

Running `pytest`: on the good test:

```bash
pytest test-foo.py::test_plan_success
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.10.5, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/ludomagno/Desktop/dev/tf-tests
collected 1 item                                                                                                                                                                             

test-foo.py .                                                                                                                                                                          [100%]

===================================================================================== 1 passed in 0.06s ======================================================================================
```

Runing `pytest` on the "bad" test:

```bash
------------------------------------------------------------------------------------- Captured log call --------------------------------------------------------------------------------------
CRITICAL tftest:tftest.py:551 Error running command plan: 1  
Error: Invalid value for variable

  on main.tf line 1:
   1: variable "foo" {
    ├────────────────
    │ var.foo is "bar"

Foo neds to be at least 3 characters.

This was checked by the validation rule at main.tf:3,3-13.
================================================================================== short test summary info ===================================================================================
FAILED test-foo.py::test_plan_fail - tftest.TerraformTestError: Error running command plan: 1  
===================================================================================== 1 failed in 0.12s ======================================================================================
```

--- a/test/test_sample_plan_error.py
+++ b/test/test_sample_plan_error.py
@@ -1,0 +1,30 @@
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Test command out attribute of tftest exception class."
+
+import pytest
+import tftest
+
+
+def test_cmd_err(fixtures_dir):
+  with pytest.raises(tftest.TerraformTestError) as e:
+    tf = tftest.TerraformTest('foobar')
+    tf.setup()
+  assert not e.value.cmd_error
+  tf = tftest.TerraformTest('plan', fixtures_dir)
+  tf.setup()
+  with pytest.raises(tftest.TerraformTestError) as e:
+    tf.plan()
+  assert e.value.cmd_error

--- a/tftest.py
+++ b/tftest.py
@@ -54,7 +54,10 @@ TerraformStateResource = collections.namedtuple(
 
 
 class TerraformTestError(Exception):
-  pass
+
+  @property
+  def cmd_error(self):
+    return self.args[1] if len(self.args) > 1 else None
 
 
 _TG_BOOL_ARGS = [
@@ -338,6 +341,8 @@ class TerraformTest(object):
               f'Unable to restore terraform file {bkp_file.resolve()}')
           raise TerraformTestError(
               f'Restore of terraform file ({bkp_file.resolve()}) failed')
+        else:
+          bkp_file.unlink(True)
 
   def _abspath(self, path):
     """Make relative path absolute from base dir."""
@@ -384,7 +389,6 @@ class TerraformTest(object):
           with open(tf_file, 'r') as src:
             terraform = src.read()
           with open(tf_file, 'w') as src:
-
             terraform = re.sub(
                 r'prevent_destroy\s+=\s+true', 'prevent_destroy = false', terraform)
             src.write(terraform)
@@ -414,7 +418,8 @@ class TerraformTest(object):
       else:
         _LOGGER.warning('no such file {}'.format(link_src))
     self._finalizer = weakref.finalize(
-        self, self._cleanup, self.tfdir, filenames, deep=cleanup_on_exit, restore_files=disable_prevent_destroy)
+        self, self._cleanup, self.tfdir, filenames, deep=cleanup_on_exit,
+        restore_files=disable_prevent_destroy)
     setup_output = self.init(plugin_dir=plugin_dir,
                              init_vars=init_vars, backend=backend, **kw)
     if workspace_name:
@@ -563,7 +568,7 @@ class TerraformTest(object):
       message = 'Error running command {command}: {retcode} {out} {err}'.format(
           command=cmd, retcode=retcode, out=full_output, err=err)
       _LOGGER.critical(message)
-      raise TerraformTestError(message)
+      raise TerraformTestError(message, err)
     return TerraformCommandOutput(retcode, full_output, err)
 
   def _tg_ra(self) -> List[str]:

